_source/_posts/2025-11-04-user-entitlements-workshop.md

```diff
@@ -0,0 +1,1298 @@
+
+In the IDE, open `okta-enterprise-ready-workshops/apps/api/src/scim-types.ts`. 
+
+Add new schema URNs after the `SCHEMA_OKTA_ROLE` definition towards the end of the file:
+
+```ts
+export const SCHEMA_OKTA_ENTITLEMENT = 'urn:okta:scim:schemas:core:1.0:Entitlement';
+export const SCHEMA_CHARACTERISTIC = 'urn:bestapps:scim:schemas:extension:todoapp:1.0:Characteristic';
+```
+
+We defined a new schema URN for the characteristic SCIM resource. Following naming conventions for extension schemas, we substituted our company name (Best Apps) and added the app's name (Todo app). The format looks like this
+
+```
+urn:<Company name>:scim:schemas:extension:<App name>:1.0:<Custom entitlement>
+```
+
+Right now, there's a custom TypeScript type for `RESOURCE_TYPES`. Since we'll have custom schemas as a resource type, update the code.
+
+```ts
+export type RESOURCE_TYPES = 'Role' | 'ResourceType' | 'Schema';
+```
+
+SCIM defines required and optional attributes to describe a schema resource. We'll define the interfaces for a schema resource. Add the following interfaces to the `scim-types.ts` file. I added mine after the other interfaces and before the URNs.
+
+```ts
+export interface ISchema {
+  id: string;
+  name?: string;
+  description?: string;
+  attributes: IAttribute[];
+  meta: IMetadata;
+}
+
+export interface IAttribute {
+  name: string;
+  description: string;
+  type: string;
+  multiValued: boolean;
+  required: boolean;
+  caseExact: boolean;
+  mutability: string;
+  returned: string;
+  uniqueness: string;
+}
+```
+
+Characteristic is a unique resource type because it's a new, custom type extending from an existing schema. We must explicitly show this relationship for consuming SCIM clients, like Okta. Find the `IResourceType` interface. We'll add a new optional property, `schemaExtensions` and inline the type definition.
+
+```ts
+export interface IResourceType {
+  id?: string;
+  schemas: string[];
+  name: string;
+  description?: string;
+  endpoint: string;
+  schema: string;
+  schemaExtensions?: {schema: string, required: boolean}[];
+  meta: IMetadata;
+}
+```
+
+SCIM clients expect a list of schemas that you offer in the SCIM server. You might've guessed what that means. You must wrap all the schemas in a SCIM `ListResponse`. Find `IListResponse` and add `ISchema` as a supported type. The `IListResponse` interface changes to:
+
+```ts
+export interface IListResponse<T extends IScimResource | IResourceType | ISchema> {
+  schemas: string[];
+  totalResults: number;
+  startIndex: number;
+  itemsPerPage: number;
+  Resources: T[];
+}
+```
+
+Finally, we define what a characteristic attribute looks like by adding the interface shown below.
```

Comment:
I am confused by calling it a "characteristic attribute" here and then having it extend IScimResource instead of IAttribute. potential rephrase: 
 
```suggestion
Finally, we define what a `characteristic` looks like by adding the interface shown below.
```